src/team_lib/mhba/xccl_mhba_lib.c

```diff
@@ -583,46 +608,52 @@ xccl_mhba_team_destroy(xccl_tl_team_t *team)
         if (status!=XCCL_OK){
             xccl_mhba_error("failed to destroy Mkeys");
         }
+        free(mhba_team->net.rkeys);
     }
     free(team);
     return status;
 }
 
-xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t *team, int fanin_value, int root)
+xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t *team, xccl_mhba_coll_req_t *request)
 {
     int i;
     int *ctrl_v;
-    if (team->node.sbgp->group_rank != root) {
-        ctrl_v = (int*)team->node.my_ctrl;
-        int v = *ctrl_v;
-        __sync_fetch_and_add(ctrl_v, (v+fanin_value));
+    int index = seq_index(request->seq_num);
+
+    if(team->occupied_operations_slots[index]){
+        return XCCL_INPROGRESS;
+    } //wait for slot to be open
+    team->occupied_operations_slots[seq_index(team->sequence_number)] = 1;
+    xccl_mhba_update_mkeys_entries(&team->node, request); // no option for failure status
+
+    if (team->node.sbgp->group_rank != request->asr_rank) {
+        *team->node.operations[index].my_ctrl = request->seq_num;
     } else {
         for (i=0; i<team->node.sbgp->group_size; i++) {
             if (i == team->node.sbgp->group_rank) {
                 continue;
             }
-            ctrl_v = (int*)((ptrdiff_t)team->node.ctrl + MHBA_CTRL_SIZE*i);
-            if (*ctrl_v != fanin_value) {
+            ctrl_v = (int*)((ptrdiff_t)team->node.operations[index].ctrl + MHBA_CTRL_SIZE*i);
+            if (*ctrl_v != request->seq_num) {
```

Comment:
the we should have (*ctrl_v < request->seq_num) instead of "!=". It is more safe, in the case if some rank performs several non-blocking alltoalls and increases the ctrl too the seq_num + 2.

Reply:
i disagree - this is per slots, so it's need to be a case where the difference isn't 2 operations, but possibly 50 operations (depend on NUM_CONCURNT_OPERATIONS).
Also, and more important, it will be indication for a bug - the fanin is for in-node sync regarding mr registration and update, and the seq num should be equal - otherwise it won't work

Reply:
ok